dotnet-engine/Yggdrasil.Engine/NativeLoader.cs

```diff
@@ -0,0 +1,179 @@
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+internal static class NativeLibLoader
+{
+    internal static IntPtr LoadNativeLibrary()
+    {
+        var libName = GetBinaryName();
+        var tempPath = Path.Combine(Path.GetTempPath(), libName);
+        var assembly = Assembly.GetExecutingAssembly();
+        var assemblyName = assembly.GetName().Name;
+
+        if (!File.Exists(tempPath))
+        {
+            using (var stream = Assembly.GetExecutingAssembly().GetManifestResourceStream($"{assemblyName}.{libName}"))
+            {
+                if (stream == null)
+                    throw new FileNotFoundException($"Embedded resource {libName} not found.");
+
+                using (var fileStream = new FileStream(tempPath, FileMode.Create, FileAccess.Write))
+                {
+                    stream.CopyTo(fileStream);
+                }
+            }
+        }
+
+        return LoadBinary(tempPath);
+    }
+
+    private static string GetBinaryName()
+    {
+        string os, arch, libc = "";
+
+        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            os = "win";
+        else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            os = "linux";
+        else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            os = "osx";
+        else
+            throw new PlatformNotSupportedException("Unsupported OS");
+
+        if (RuntimeInformation.ProcessArchitecture == Architecture.X64)
+            arch = "x86_64";
+        else if (RuntimeInformation.ProcessArchitecture == Architecture.Arm64)
+            arch = "arm64";
+        else if (RuntimeInformation.ProcessArchitecture == Architecture.X86)
+            arch = IntPtr.Size == 4 ? "i686" : "x86_64";
```

Comment:
We can't actually ask the OS if we're running on a 32 bit system because Windows will let you emulate a 32 bit system on a 64 bit arch. Soooo... we check the size of a pointer, which necessarily matches the word size

Reply:
Ok, so `Architecture.X64` and `Architecture.X86` is the same thing when pointer size is 8?

Reply:
Yep